### ISSUE TITLE
Do we still recommend the '~/Documents/basiljs/user' folder?

### ISSUE THREAD

**Issue by fabianmoronzirfas** [basiljs/basil.js#155]
As the title says. I would love to go ahead and change the installers so the question is:  

Do we still need this folder? I don't think so. Basil is now portable and the user should be able to place all his files wherever he wants them to be. Enforcing the user folder would also mean that all changes made by @trych in PR #150 are not right 🤓

**Comment by b-g** [comment]
I'm happy to change this recommendation with 2.0 ... my blueprint would be Processing e.g.

```
~/Documents/Processing3/examples
~/Documents/Processing3/libraries
~/Documents/Processing3/modes
~/Documents/Processing3/templates
~/Documents/Processing3/tools
```

**Comment by trych** [comment]
What do you mean by installers? The download bundle? Or the docs?

What is your idea of the workflow, where would you put the basil.js file? I personally will still one central spot (and for this `~Documents/basiljs` makes sense) and not have a copy of basil.js for each script that I write.

No, we don't need really need the folder anymore, however I feel it could be helpful to at least suggest one location where you could put basil.js, so inexperienced users have a guide what to do. The problem is, the example scripts need to point to _some_ location, so I thought pointing it to `~Documents/basiljs/basil.js` makes more sense. If there are better alternatives, I am happy to make another PR to change the paths again. :)

**Comment by b-g** [comment]
+1 IMO we should recommend a central location to help inexperienced users:

```
~/Documents/basiljs/basil.js
~/Documents/basiljs/sketches/
~/Documents/basiljs/examples/ ...
~/Documents/basiljs/libraries/ ... where to put e.g. underscore.js
```

**Comment by fabianmoronzirfas** [comment]
I currently not talking about the location of Basil itself. I mean the user folder. If we have Basil at `~/Documents/basiljs/basil.js` there is no room for the user folder. Telling the inexperienced user to just replace a part of that folder will create errors. That is why the `user` folder next to the `bundle` folder was introduced.   

@b-g how to update basil.js in this structure?

@trych The `install.bat` and `install.command` scripts

**Comment by b-g** [comment]
I would tell them to put all their "stuff" into `sketches` and get rid of the `user` folder.

**Comment by fabianmoronzirfas** [comment]
So the update process would be:  

1. Download latest version
2. move `sketches` out of `~/Documents/basiljs`
3. replace the whole folder `~/Documents/basiljs` 
4. move sketches folder back in

---

Or 
1. Download just basil.js file
2. replace basil.js file in `~/Documents/basiljs`

In this case we need to pack the release somehow different as github does. It only should contain:   

- basil.js
- examples/*
- sketches/.gitkeep
- libraries/.gitkeep

**Comment by trych** [comment]
In the Skype call we decided that the example will be a separate download. Or am I confusing things here?

**Comment by fabianmoronzirfas** [comment]
@trych a yeah I remember. Yes that makes it ~~(even more confusing)~~ easier.  

So we have 3 different processes.  

1. Fresh install (should contain pre defined folders and examples)
2. update basil.js (bundled file only)
3. update examples (overwrite examples folder in basil location)

is that right?  
(Having a own IDE like Processing does makes so many things easier. In P5 3.x the user does not have to touch anything anymore)

**Comment by trych** [comment]
Sounds right, yes.

What about the link between InDesign's script panel and the sketches folder? That was in the past always by far the most confusing part to set up for new users. Is there any way to set this up automatically.

**Comment by fabianmoronzirfas** [comment]
Problematic. We don't know what kind of ID Version the user is on. 

A quick search came up with these threads.  

http://stackoverflow.com/questions/9296896/how-to-create-a-mac-os-x-alias-kind-file-in-terminal

http://apple.stackexchange.com/questions/2991/whats-the-difference-between-alias-and-link

http://apple.stackexchange.com/questions/51709/can-i-create-a-desktop-shortcut-alias-to-a-folder-from-the-terminal

tl;dr

We can use symlinks or finder aliases for that on macOS. The symlink is created from terminal like this:  

    ln -s /Users/foo/absolut/path/to/examples /absolut/path/to/target/InDesign/scripts\ panel 

The Alias needs some [osascript/AppleScript magic](http://apple.stackexchange.com/a/51712/52621) to be created.  

The only downside to symlinks is that they don't get indexed my spotlight (as far as I can read from these links).  

I have no idea how to do that on Windows.

**Comment by trych** [comment]
> In this case we need to pack the release somehow different as github does. It only should contain:
> 
>     basil.js
>     examples/*
>     sketches/.gitkeep
>     libraries/.gitkeep

This also means that we'll have to add these three folders to the `.gitignore` file, right?
The reason is that the standard location of the basil.js file change from `~/Documents/basiljs/bundle/basil.js` to `~/Documents/basiljs/basil.js` which means that also our git repository for development has to move one level up. Therefore we would need to exclude those three folders that they don't become part of the git repo, but are only placed there (via install script or whatnot) by the end user.

Is this correct? If so, I can go ahead for a quick PR with the `.gitignore` file.

**Comment by fabianmoronzirfas** [comment]
I was thinking of creating a npm script that creates the download bundle and that should be added to the page as the official download. So the enduser never sees the development folders and all our package.json files and so on.

**Comment by trych** [comment]
Yes, of course the end user should not see any development folders. I am rather talking about our situation as developers. We will need to have both the development environment as well as the end user environment set up simultaneously.

As we have changed basil's standard path (for example in the example scripts) we also need to move our development environment from `~/Documents/basiljs/bundle` to `~/Documents/basiljs/`. And as this is the location where the new `sketches`, `libraries` and `examples` folder reside, we should add those to `.gitignore`.

I mean, I can at the moment avoid conflicts by adding them to `.git/info/exlude` for my own environment, but they should actually be added to the `.gitignore` itself to allow all developers to have both environments at the same time.

**Comment by b-g** [comment]
Hi! Happy new '17!

Yes I remember now ... this was one of the main reasons why we went down the `~/Documents/basiljs/bundle` route :) But still I find the proposed new structure better. So I guess it would be something like (correct me if I'm wrong):

install.command:
* moves (and overwrites) `basil.js`-> `~/Documents/basiljs/basil.js`
* moves (and overwrites) `examples`-> `~/Documents/basiljs/examples`
* moves (and overwrites) `docs`-> `~/Documents/basiljs/examples`
* creates folder (if not existing, no overwrite)  `~/Documents/basiljs/sketches`
* creates folder (if not existing, no overwrite)  `~/Documents/basiljs/libraries`

.gitignore:
sketches/
libraries/

**Comment by trych** [comment]
Happy 2017! 🎉 

> moves (and overwrites) docs-> ~/Documents/basiljs/examples

You probably mean `~/Documents/basiljs/docs`, right? But do we have a `docs` folder? Or do you mean something different altogether?

For new developers we should probably add a description to the Wiki about how to set up the developing environment, because a simple git clone won't work now (as we have to preserve or newly create the `sketches/` and `libraries/` folder). Or maybe Fabian knows of some npm magic of how to automate the setup. 😃

Will do a PR for the minor .gitignore changes.

**Comment by b-g** [comment]
@trych Sorry for the confusion ... yes I meant a `docs` folder to hold a potential static offline html api documentation. (Can be very handy to have the docs locally on your machine especially in workshop situations with shaky WIFI connections. But IMO tbd once we know how the docs are generated, could also be a separate download)

**Comment by b-g** [comment]
Will close this issue now, as we all agreed on the new folder structure. (Plz re-open if I'm wrong :)
